### ISSUE TITLE
Empty arrays don't pass validate exclusion of nil in Rails 6.1

### ISSUE THREAD

**Issue by haines** [rails/rails#41051]
We'd like to validate that an attribute backed by an array or JSON(B) database column is not nil, while allowing empty arrays.

Before Rails 6.1 `validates :value, exclusion: { in: [nil] }` worked.

After https://github.com/rails/rails/pull/22610 this is no longer the case.

Also reported by @mkamensky here: https://github.com/rails/rails/pull/22610#commitcomment-45326313

### Steps to reproduce

Attempt to set an attribute with `validates :value, exclusion: { in: [nil] }` to `[]`.

[Repro script](https://gist.github.com/haines/90d8abb93c7ca00279a723e20c32e072) (requires PostgreSQL running locally)

```console
$ git clone https://gist.github.com/90d8abb93c7ca00279a723e20c32e072.git validates_exclusion_of_nil_with_empty_array

$ cd validates_exclusion_of_nil_with_empty_array

$ ruby validates_exclusion_of_nil_with_empty_array_test.rb 6.0.3.4
-- create_table(:examples)
   -> 0.0196s
Run options: --seed 50787

# Running:

..

Finished in 0.016144s, 123.8850 runs/s, 123.8850 assertions/s.
2 runs, 2 assertions, 0 failures, 0 errors, 0 skips

$ ruby validates_exclusion_of_nil_with_empty_array_test.rb 6.1.1
-- create_table(:examples)
   -> 0.0199s
Run options: --seed 6233

# Running:

F

Failure:
TestValidateExclusionOfNilWithEmptyArray#test_empty_array_is_valid [validates_exclusion_of_nil_with_empty_array_test.rb:39]:
Expected false to be truthy.


rails test validates_exclusion_of_nil_with_empty_array_test.rb:38

.

Finished in 0.006463s, 309.4538 runs/s, 309.4538 assertions/s.
2 runs, 2 assertions, 1 failures, 0 errors, 0 skips
```

### Expected behavior

`validates :value, exclusion: { in: [nil] }` should allow `[]` but reject `nil`

### Actual behavior

`validates :value, exclusion: { in: [nil] }` rejects both `[]` and `nil`

### System configuration
**Rails version**: 6.1.1

**Ruby version**: 2.7.2p137

**Comment by rafaelfranca** [comment]
Good point. Since you know how to reproduce, can you open a PR fixing the behavior? Adding a new option to the behavior of #22610 and revert the behevior without that option set to the old one would do it.

**Comment by haines** [comment]
Would it make sense for that option to be available for _all_ validations?

I think it's generally useful to be able to apply a validation to every value in an array column; I don't see why it should be limited to in/exclusivity.

**Comment by mkamensky** [comment]
Please note my other comment https://github.com/rails/rails/pull/22610#issuecomment-749246114, which I think is a more serious issue

**Comment by intrip** [comment]
@haines if you are no longer interested I'd be happy to work on a fix.

**Comment by haines** [comment]
Hey @intrip, happy for you to push this forward. I do think it would be better for this to be a generic option for all validators (apply this validation to all values in the array) rather than a special case for *clusion validators.

**Comment by rafaelfranca** [comment]
Let's start with the small scope since it will fix a real problem. Later, if you believe we can add this option to all association we can review and discuss the merits of that addition separately.

**Comment by intrip** [comment]
@rafaelfranca I've created #41412 which implements [your suggestion](https://github.com/rails/rails/issues/41051#issuecomment-757037953).

The idea behind the `all` option name is outlined [here](https://github.com/rails/rails/pull/41412#discussion_r575112561).
If you agree I could create a follow up PR to implement also the `any` option which makes sense for the `exclusion` validator.

**Comment by intrip** [comment]
not stale, we have #41412 which should close this

**Comment by moznion** [comment]
Is there any progress on this issue and related pull request #41412? I’ve just made a quick patch — how about using this as a temporary solution?

```diff
diff --git a/activemodel/lib/active_model/validations/clusivity.rb b/activemodel/lib/active_model/validations/clusivity.rb
index 125c85609d..0a4afd3412 100644
--- a/activemodel/lib/active_model/validations/clusivity.rb
+++ b/activemodel/lib/active_model/validations/clusivity.rb
@@ -22,6 +22,11 @@ def include?(record, value)
         members = resolve_value(record, delimiter)

         if [http://value.is](https://t.co/phE8uWSmHd)_a?(Array)
+          # If value is an empty array, it validates incorrectly because `value.all?` always returns true.
+          # Therefore, this handles the empty array as a special case.
+          if value.empty?
+            return members.public_send(inclusion_method(members), value)
+          end
           value.all? { |v| members.public_send(inclusion_method(members), v) }
         else
           members.public_send(inclusion_method(members), value)
```
